### ISSUE TITLE
Question about fully trained DeepPhys model weights

### ISSUE THREAD

**Issue by Bel0rusochka** [ubicomplab/rPPG-Toolbox#372]
Hello,

I am currently working on my bachelor's thesis on deepfake detection and would like to use the DeepPhys model as the foundation for my method. However, I noticed that the weights I found in the repository do not seem to work correctly. I am using BP4D_PseudoLabel_DeepPhys, but when applied, the model produces strange results.

I wanted to ask:

1. How well is this model trained? Could it be just an example trained on a small dataset?
2. Do you have a fully trained version of the DeepPhys model with weights that can be used in real experiments?

I would appreciate any information. Thank you!

**Comment by yahskapar** [comment]
Hi @Bel0rusochka,

Not sure regarding your observation on DeepPhys performance on BP4D in particular - maybe @girishvn can comment since he added that dataset. You can also check out the [BigSmall paper](https://arxiv.org/abs/2303.11573) from which that pre-trained DeepPhys model came from. 

Is there a particular reason you're exploring DeepPhys in particular? You could try other pre-trained models in the `final_model_release` folder of this repo, perhaps the one trained on motion augmented UBFC-rPPG data [here](https://github.com/ubicomplab/rPPG-Toolbox/blob/main/final_model_release/MA-UBFC_deepphys.pth), but I'm not sure how helpful those other models would be depending on your test data.

It's worth noting DeepPhys is a fairly old model that has been beaten in performance by numerous models (e.g., [TS-CAN](https://proceedings.neurips.cc/paper_files/paper/2020/file/e1228be46de6a0234ac22ded31417bc7-Paper.pdf), [BigSmall](https://arxiv.org/abs/2303.11573), and [PhysMamba](https://arxiv.org/pdf/2409.12031)) that are also available in this toolbox. Some very recent papers and their corresponding methods (see #335 and #339) will also be added to this toolbox in the coming few weeks. Furthermore, there are techniques such as [motion augmentation](https://motion-matters.github.io/) which may help improve performance depending on the kind of test data you have in mind (perhaps highly unconstrained self-captured videos).

**Comment by girishvn** [comment]
Hi @Bel0rusochka, 

Can you explain a little more on the poor performance you are seen for the BP4D trained DeepPhys model (e.g. what data are you validating on, etc.). The saved model is the same model that was used for Table 3 of the paper: https://arxiv.org/pdf/2210.00716.

If you are testing with some different data there are a few considerations to take into account when working w/ BP4D trained models.

1. The BP4D dataset only contains stationary videos of ppl (no motion artifacts common in other datasets).
2. The BP4D dataset has blank / black background which can cause models trained on the dataset to generalize poorly to dataset with noisy background environments.
3. The BP4D dataset does NOT have PPG labels. For this reason models trained on BP4D use a derived pseudo label as described in our paper and this repo's ReadMe. 

Any / all of these may contribute to unexpected performance.
As yahskapar mentioned, there are several other models which have surpassed BP4D over the years (e.g. TSCAN, etc.). I would suggest using a different model architecture, or you can also choose a DeepPhys model trained on a different dataset (e.g. PURE).

Better yet, I would try to use a model trained with motion augmentations (e.g. this MA UBFC model: https://github.com/ubicomplab/rPPG-Toolbox/blob/main/final_model_release/MA-UBFC_deepphys.pth).

Hope this helps, and best of luck with your bachelor's thesis!

**Comment by Bel0rusochka** [comment]
Hi @girishvn and @yahskapar,
The extracted rPPG signal from DeepPhys model values were close to zero, with only small fluctuations not exceeding 0.25, which seemed unusual. I tested the model on videos from the UBFC-rPPG dataset to compare the extracted signals with ground truth data.

I also tried a few other pre-trained DeepPhys models trained on different datasets, but I didn’t observe significant improvements. However, when I switched to a pre-trained PhysNet model, I get much better results. 

It can be stated that the problem is solved. Thank you for your help!

**Comment by girishvn** [comment]
Hi @Bel0rusochka,

Glad you found something that worked. 
A quick think to note is that the output dynamic range of each model varies (e.g. the normal output range of values).
As many ML models experience more stable trainer with smaller outputs it is possible that this is normal behavior.

What you can try to validate this is to normalize the signal by the max value (e.g. re-map the signal so the map value is 1).
You may observe more expected behavior in the visualization of the signal.

It should be noted that most reported metrics are in terms of HR accuracy / error, and thus do no account for the magnitude / morphology of the output waveform, just its frequency.
